### PR TITLE
Id map services

### PR DESCRIPTION
**Pull request by lucas-ubm**

### CODE REVIEW

File: docs/content/recipes/interoperability/identifier-mapping.md [hunk 1]
```diff
@@ -0,0 +1,384 @@
...
+
+The main goals of this recipe are:
+
+> - To understand where the need arises for mapping between data identifiers;
+> - How to represent equivalences for use by others; and
+> - Know what services are available to support data interoperability.
+
+This recipe assumes that you are already familiar with identifiers and the minting of identifiers.
+
+*For more details on identifiers, see the [How to generate globally unique, resolvable and persistent identifiers](https://fairplus.github.io/cookbook-dev/recipes/findability/identifiers.html) recipe.*
```
Comment (egonw) [inline docs/content/recipes/interoperability/identifier-mapping.md:79]: it may be possible to have the link to `identifiers.md` instead of `identifiers.html`. Several tools convert the link automatically when the HTML is generated, and this approach has the advantage that you can then use a relative link so that you can follow it on GitHub in the development and/or main branch too.

File: docs/content/recipes/interoperability/identifier-mapping.md [hunk 2]
```diff
@@ -0,0 +1,384 @@
...
+      I[Identifier]
+      I --> IM[Identifier Mapping]
+      M[Mapping] --> OM[Ontology Mapping]
+      OM --> OxO
+
+      M --> IM
+
+      IM -- has section --> IMS
+
+      IM -- has section--> IE([Identifier Equivalence])
```
Comment (egonw) [inline docs/content/recipes/interoperability/identifier-mapping.md:96]: Check consistent whitespace around label of the edge. Maybe not technically important, but makes it easier to read.

File: docs/content/recipes/interoperability/identifier-mapping.md [hunk 3]
```diff
@@ -0,0 +1,384 @@
...
+       class M select;
+       class IE select;
+       class SL select;
+      %% style IM fill:#f9f
+      %% style I fill:#f9f
+      %% style OxO fill:#f9f
+```
+___
+
+**ToDo:** Update diagram
```
Comment (egonw) [inline docs/content/recipes/interoperability/identifier-mapping.md:132]: Check.

File: docs/content/recipes/interoperability/identifier-mapping.md [hunk 4]
```diff
@@ -0,0 +1,384 @@
...
+
+| Data Formats  | Terminologies | Models  |
+| :------------- | :------------- | :------------- |
+| [IRI](https://tools.ietf.org/html/rfc3987) |   |   |
+| [CURIE](https://www.w3.org/TR/2010/NOTE-curie-20101216/) |   |   |
+| [URL](https://tools.ietf.org/html/rfc1738) |  |  |
+| [HTTP]() |  |  |
+| [RDF]() |  |  |
+| [CSV]() |  |  |
+| [TSV]() |  |  |
```
Comment (egonw) [inline docs/content/recipes/interoperability/identifier-mapping.md:177]: Four missing links.

File: docs/content/recipes/interoperability/identifier-mapping.md [hunk 5]
```diff
@@ -0,0 +1,384 @@
...
+| [VoID Linkset]() |  |  |
+
+---
+## Mappings
+Before diving into identifier mapping it is important to understand the possible types of mappings that can be performed between entities. While initially we might think of mapping as simply linking identical entities in different databases/formats, sometimes related entities might also be of interest. 
+When these mappings might only be interesting depending on the context in which data is being used, we run into a situation that has been described as "scientific lenses". These lenses allow us
+to dynamically select which mappings to consider relevant and which to ignore. For example allowing or disallowing mappings between stereoisomers or between genes and proteins.
+
+Examples of types of mappings are:
+* **Content mapping**: where we are mapping the actual entities by using techniques such as BLAST in biological sequences or comparison of INCHI identifiers for chemical compounds
```
Comment (egonw) [inline docs/content/recipes/interoperability/identifier-mapping.md:187]: INCHI -> InChI

File: docs/content/recipes/interoperability/identifier-mapping.md [hunk 6]
```diff
@@ -0,0 +1,384 @@
...
+| [HTTP]() |  |  |
+| [RDF]() |  |  |
+| [CSV]() |  |  |
+| [TSV]() |  |  |
+| [VoID Linkset]() |  |  |
+
+---
+## Mappings
+Before diving into identifier mapping it is important to understand the possible types of mappings that can be performed between entities. While initially we might think of mapping as simply linking identical entities in different databases/formats, sometimes related entities might also be of interest. 
+When these mappings might only be interesting depending on the context in which data is being used, we run into a situation that has been described as "scientific lenses". These lenses allow us
```
Comment (egonw) [inline docs/content/recipes/interoperability/identifier-mapping.md:183]: Add reference to one of the papers.

File: docs/content/recipes/interoperability/identifier-mapping.md [hunk 7]
```diff
@@ -0,0 +1,384 @@
...
+to dynamically select which mappings to consider relevant and which to ignore. For example allowing or disallowing mappings between stereoisomers or between genes and proteins.
+
+Examples of types of mappings are:
+* **Content mapping**: where we are mapping the actual entities by using techniques such as BLAST in biological sequences or comparison of INCHI identifiers for chemical compounds
+* **Ontology mapping**: this can either be 
+    * As a direct 1-to-1 mapping between equivalent terms in different ontologies
+    * As a complex m-to-m mapping between terms in different ontologies taking into account their hierarchical structure, see [[2]](#References).
+* **Identifier mapping**: The focus of this relationship. This can either be:
+    * Mapping between differently formed identifiers that resolve to the same entity.
```
Comment (egonw) [inline docs/content/recipes/interoperability/identifier-mapping.md:192]: add an example here

File: docs/content/recipes/interoperability/identifier-mapping.md [hunk 8]
```diff
@@ -0,0 +1,384 @@
...
+
+Databases often contain data that exists in, or is closely related to, the content of other databases. For example, the genome database [Ensembl](http://ensembl.org/) contains data about genes that are related to entries in databases such as HUGO Gene Nomenclature Committee (https://www.genenames.org/) or [NCBI Gene](https://www.ncbi.nlm.nih.gov/gene/); or a database about drugs, e.g. [DrugBank](https://drugbank.com/), often contains details of the chemical substance that forms the drug which are also contained in chemical databases such as [ChEBI](https://www.ebi.ac.uk/chebi/) or [PubChem](https://pubchem.ncbi.nlm.nih.gov/). This results in a large number of unique identifiers notionally for the same concept. This large number of identifiers for the same concept prevents interoperability of the data, since additional knowledge is needed to know which identifiers represent the same concept from different databases. However, the databases often contain `cross-reference` links to other databases to represent these `equivalences`.
```
Comment (egonw) [inline docs/content/recipes/interoperability/identifier-mapping.md:206]: Consider adding proper literature citations in addition to the URLs.

File: docs/content/recipes/interoperability/identifier-mapping.md [hunk 9]
```diff
@@ -0,0 +1,384 @@
...
+
+The simplest way to exchange equivalences is in a simple text file, which could be structured as a tab-separated-value (TSV) file. Such a file usually consists of two columns, one per dataset, and each row represents an equivalence declaration. The interpretation is that the two identifiers on the same row are equivalent in some way. These files tend to carry little to no metadata about the mappings, i.e. the mechanism by which the mapping was derived is not given, nor are details of the verison of the datasets that were linked.
+
+The following example shows the mapping equivalences between ChEMBL target components (proteins) and UniProt proteins.
+
+```
+ChEMBL_Target_Component    UniProt
+CHEMBL_TC_4803    A0ZX81
+CHEMBL_TC_2584    A1ZA98 
+```
```
Comment (egonw) [inline docs/content/recipes/interoperability/identifier-mapping.md:234]: Is this mean to be a bit of TSV content? then use TABs and use the TSV syntax highlighting.

File: docs/content/recipes/interoperability/identifier-mapping.md [hunk 10]
```diff
@@ -0,0 +1,384 @@
...
+
+
+#### Simple Standard for Sharing Ontology Mappings (SSSOM)
+
+The OBOFoundry Simple Standard for Sharing Ontology Mappings ([SSSOM](https://github.com/OBOFoundry/SSSOM)) is a newly emerging standard for exchanging mapping information with minimal metadata, although the minimal model is extensible. It consists of a tab-separated-value file (TSV) with each row representing a mapping. The columns in the file have been defined by the community ([latest version](https://github.com/OBOFoundry/SSSOM/blob/master/SSSOM.md)) and provide the mapping together with its provenance. At present, four columns are required (`subject_id`, `predicate_id`, `object_id`, `match_type`) with optional columns to provide more provenance to support the mapping, e.g. licensing information, author information, creation method. The use of CURIEs within the TSV is strongly encouraged.
+
+The following TSV shows our example data as a mapping file using the minimal columns (correct as of November 2020). The information provided is less than the minimal VoID model above.
+
+```
```
Comment (egonw) [inline docs/content/recipes/interoperability/identifier-mapping.md:273]: Use TSV syntax highlighting

File: docs/content/recipes/interoperability/identifier-mapping.md [hunk 11]
```diff
@@ -0,0 +1,384 @@
...
+subject_id    predicate_id    object_id    match_type
```
Comment (egonw) [inline docs/content/recipes/interoperability/identifier-mapping.md:274]: Double check use of TABs, because they seem to be spaces

File: docs/content/recipes/interoperability/identifier-mapping.md [hunk 12]
```diff
@@ -0,0 +1,384 @@
...
+
+### Identifier Mapping services
+
+Identifier mapping services are databases that contain lists of identifiers, often from different databases, that are known to be equivalent. These are consumed from original data sources and third parties in one of the formats provided above.
+
+The common functionality offered by these services is to return a set of equivalent identifiers for a given identifier. That is, you can ask these services for all identifiers that are equivalent to your identifier. The services will vary in the response depending on their coverage of data sources and whether they decide to compute the transitive closure of the identifiers. That is, if data source A declares that `A:id1` is equivalent to `B:acc32`, and data source C declares that `C:67cb2865-781f-4450-a99c-e9b33bf4d5b5` is equivalent to `B:acc32`, should a lookup for equivalent identifiers for `A:id1` return `C:67cb2865-781f-4450-a99c-e9b33bf4d5b5` since `A:id1 <=> B:acc32` and `B:acc32 <=> C:67cb2865-781f-4450-a99c-e9b33bf4d5b5`.
+
+The following is an incomplete list of identifier mapping services.
+
+* [bridgedb.org](https://bridgedb.org/)
```
Comment (egonw) [inline docs/content/recipes/interoperability/identifier-mapping.md:295]: Bc of still hard to solve httpS issues, please use https://bridgedb.github.io/

File: docs/content/recipes/interoperability/identifier-mapping.md [hunk 13]
```diff
@@ -0,0 +1,384 @@
...
+
+    > [BridgeDb](https://bridgedb.org/) is a framework for identifier mapping within the life sciences which covers genes, proteins, genetic variants, metabolites, and metabolic reactions. It is provided as a web service, a standalone application that can be installed locally, a Java library or an R Package.
```
Comment (egonw) [inline docs/content/recipes/interoperability/identifier-mapping.md:297]: https://bridgedb.github.io/

Comment (egonw) [reply docs/content/recipes/interoperability/identifier-mapping.md:297]: Oh, and add the citation to the paper.

File: docs/content/recipes/interoperability/identifier-mapping.md [hunk 14]
```diff
@@ -0,0 +1,384 @@
...
+    > 
+    > It permits users to lookup equivalent database identifiers for a given database identifier within a specified organism. The following `curl` command to the REST API retrieves the equivalent identifiers for the EntrezGene (now known as NCBI Gene) `L` identifier `1234` for the `Human` gene [CCR5](https://www.ncbi.nlm.nih.gov/gene/1234) as a TSV file.
+    > ```
+    > curl -X GET "https://webservice.bridgedb.org/Human/xrefs/L/1234" -H "accept: */*"
+    > ```
+    > For more details about using BridgeDb as an identifier mapping service, please see [this recipe]XXX.
+
+* [UniChem](https://www.ebi.ac.uk/unichem/)
+    > [UniChem](https://www.ebi.ac.uk/unichem/) is a specialised identifier mapping service for chemical structures. For a chemical structure -- specified as an identifier, InChI, or InChI Key -- it will equivalent structures found in the [EMBL-EBI](https://www.ebi.ac.uk/) chemistry resources.
```
Comment (egonw) [inline docs/content/recipes/interoperability/identifier-mapping.md:306]: Add paper.

File: docs/content/recipes/interoperability/identifier-mapping.md [hunk 15]
```diff
@@ -0,0 +1,384 @@
...
+    > [sameas.org](https://sameas.org/) is a general purpose service that will return a set of equivalent URLs for a given URL. The equivalences are based on an incomplete set of `owl:sameAs` statements contained in data available on the web.
+    > 
+    > The following `curl` command retrieves the equivalent URLs for EBI RDF Platform representation of ChEMBL [DIAZEPAM](https://rdf.ebi.ac.uk/resource/chembl/molecule/CHEMBL12) as a JSON object.
+    > ```
+    > curl -iLH "Accept: application/json" "http://sameas.org/?uri=http://rdf.ebi.ac.uk/resource/chembl/molecule/CHEMBL12"
+    > ```
+    > Note that the coverage of sameas.org within the life sciences is very small.
+
+
+* [Identifiers.org](http://identifiers.org/)
```
Comment (egonw) [inline docs/content/recipes/interoperability/identifier-mapping.md:324]: httpS://identifiers.org/ seems to be the preferred link now.

**Review by egonw — CHANGES_REQUESTED** [review]
I suggest a number of smaller and larger changes. Nothing too significant. I will try to find some time to write some patches myself (but no promises).

File: docs/content/recipes/interoperability/identifier-mapping.md (hunk 1 above)
Comment (lucas-ubm) [reply docs/content/recipes/interoperability/identifier-mapping.md:79]: I think we are waiting for permanent links for the recipe before we have final links between them so whatever we use to link will probably be changed later on

File: docs/content/recipes/interoperability/identifier-mapping.md (hunk 2 above)
Comment (lucas-ubm) [reply docs/content/recipes/interoperability/identifier-mapping.md:96]: Graph updated

File: docs/content/recipes/interoperability/identifier-mapping.md (hunk 3 above)
Comment (lucas-ubm) [reply docs/content/recipes/interoperability/identifier-mapping.md:132]: Diagram deleted as discussed with @AlasdairGray

File: docs/content/recipes/interoperability/identifier-mapping.md (hunk 4 above)
Comment (lucas-ubm) [reply docs/content/recipes/interoperability/identifier-mapping.md:177]: Added some links, not sure if the correct ones, @AlasdairGray can check

File: docs/content/recipes/interoperability/identifier-mapping.md (hunk 5 above)
Comment (lucas-ubm) [reply docs/content/recipes/interoperability/identifier-mapping.md:187]: Corrected

File: docs/content/recipes/interoperability/identifier-mapping.md (hunk 6 above)
Comment (lucas-ubm) [reply docs/content/recipes/interoperability/identifier-mapping.md:183]: Added reference to paper [1] from references

File: docs/content/recipes/interoperability/identifier-mapping.md (hunk 9 above)
Comment (lucas-ubm) [reply docs/content/recipes/interoperability/identifier-mapping.md:218]: Changed it to tabs, how do you go about using TSV syntax highlighting?

I tried 
```tsv

```

but that didn't change anything

File: docs/content/recipes/interoperability/identifier-mapping.md (hunk 11 above)
Comment (lucas-ubm) [reply docs/content/recipes/interoperability/identifier-mapping.md:258]: Is this what you mean?

\```tsv
subject_id	predicate_id	object_id	match_type
chembl:CHEMBL_TC_4803	skos:exactMatch	uniprot:A0ZX81	sio:database-cross-reference
chembl:CHEMBL_TC_2584	skos:exactMatch	uniprot:A1ZA98	sio:database-cross-reference
\```

```tsv
subject_id	predicate_id	object_id	match_type
chembl:CHEMBL_TC_4803	skos:exactMatch	uniprot:A0ZX81	sio:database-cross-reference
chembl:CHEMBL_TC_2584	skos:exactMatch	uniprot:A1ZA98	sio:database-cross-reference
```

File: docs/content/recipes/interoperability/identifier-mapping.md (hunk 12 above)
Comment (lucas-ubm) [reply docs/content/recipes/interoperability/identifier-mapping.md:279]: Done

File: docs/content/recipes/interoperability/identifier-mapping.md (hunk 13 above)
Comment (lucas-ubm) [reply docs/content/recipes/interoperability/identifier-mapping.md:281]: Done

I added
[BridgeDb](https://bridgedb.github.io/)  [[2]](#References) 

And added paper to references

File: docs/content/recipes/interoperability/identifier-mapping.md (hunk 14 above)
Comment (lucas-ubm) [reply docs/content/recipes/interoperability/identifier-mapping.md:290]: Added 
Chambers, J., Davies, M., Gaulton, A., Hersey, A., Velankar, S., Petryszak, R., Hastings, J., Bellis, L., McGlinchey, S., & Overington, J. P. (2013). UniChem: A unified chemical structure cross-referencing and identifier tracking system. Journal of Cheminformatics, 5(1), 3. https://doi.org/10.1186/1758-2946-5-3

to references and cited here

File: docs/content/recipes/interoperability/identifier-mapping.md (hunk 15 above)
Comment (lucas-ubm) [reply docs/content/recipes/interoperability/identifier-mapping.md:308]: Done

File: docs/content/recipes/interoperability/identifier-mapping.md (hunk 8 above)
Comment (lucas-ubm) [reply docs/content/recipes/interoperability/identifier-mapping.md:190]: Alasdair added these links so not sure whether I should change it

File: docs/content/recipes/interoperability/identifier-mapping.md (hunk 7 above)
Comment (lucas-ubm) [reply docs/content/recipes/interoperability/identifier-mapping.md:176]: Do these examples make sense to you?

* e.g. mapping the ontology term for gene in GO and HGNC
* e.g. mapping a gene to all its possible gene products 
* e.g. mapping a gene from HGNC to the same gene in Ensembl

**Comment by proccaserra** [comment]
Hi @lucas-ubm @egonw  checking with you before merging: are you done with the edits? can I proceed with merging to dev?

no pb if you need more time.

I have set this recipe for Milestone 2 (end of February)

thx

**Comment by lucas-ubm** [comment]
Not done yet, I need to discuss a couple things with Egon and Alasdair, I will @ you when it's done (before milestone)

**Comment by AlasdairGray** [comment]
@lucas-ubm @proccaserra do we have a list of recipes that we can put in at the end of the Identifier Equivalences section?

**Comment by AlasdairGray** [comment]
@lucas-ubm @proccaserra The section on Identifier Mapping Services needs a link put in for the BridgeDb recipel

**Comment by AlasdairGray** [comment]
Other than the two comments above, I believe that this is ready to be merged in.
@proccaserra @lucas-ubm up to you now.

**Comment by lucas-ubm** [comment]
> @lucas-ubm @proccaserra The section on Identifier Mapping Services needs a link put in for the BridgeDb recipe

I am not sure how to link between recipes just yet so I wasn't sure about how to incorporate it. I thought it had to be published first and only then could links be added, is that right @proccaserra ? Otherwise how can I link?

**Comment by lucas-ubm** [comment]
After discussing with @proccaserra during the cookbook dash:

* I will be adding the references between recipes. Philippe told me how to do it, I also need to redo the way references are done because a bibtex file should be used (following Turing book guidelines)
* Regarding recipes to add at the end of the identifier equivalences section we will be waiting for a Provenance recipe to link there, meanwhile the following paragraph suggested by Philippe will be added.

For reference the current ending is 

> Depending on the nature of the data, there are different ways that equivalences can be computed. 

And the following will be added
> Some elements to take into consideration are: 
> 1. an indicator of identity (similar to what is found in the ontology world with skos:sameAs or skos:narrower) 
> 1. the evidence backing up that indication (e.g. a similarity measure, a score). 
> 1. audit&trail information (who,what,when, e.g. agent X, mapping tools Y, DDMMYYYY, ).

Does this cover your suggestions @AlasdairGray ?

**Comment by AlasdairGray** [comment]
> For reference the current ending is
> 
> > Depending on the nature of the data, there are different ways that equivalences can be computed.
> 
> And the following will be added
> 
> > Some elements to take into consideration are:
> > 
> > 1. an indicator of identity (similar to what is found in the ontology world with skos:sameAs or skos:narrower)
> > 2. the evidence backing up that indication (e.g. a similarity measure, a score).
> > 3. audit&trail information (who,what,when, e.g. agent X, mapping tools Y, DDMMYYYY, ).
> 
> Does this cover your suggestions @AlasdairGray ?

On the whole, yes, but I would rephrase to
1. a mapping predicate taken from a well known ontology, e.g. `owl:sameAs` or `skos:narrower`
2. the evidence behind the equivalence claims, e.g. a similarity score or the property on which the equivalence is based such as InChI Key for chemicals
3. audit trail information, i.e. who, what, when, e.g. agent X using mapping tool Y on YYYY-MM-DD

**Comment by robertgiessmann** [comment]
the changes were integrated via a different branch, thus I close this pull request.
